instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/AttributeBindingFactory.java

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.tracer;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.instrumentation.api.internal.ParameterizedClass;
+import java.lang.reflect.Type;
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Helper class for creating {@link AttributeBinding} instances based on the {@link Type} of the
+ * parameter for a traced method.
+ */
+class AttributeBindingFactory {
+  private AttributeBindingFactory() {}
+
+  static AttributeBinding createBinding(String name, Type type) {
+
+    // Simple scalar parameter types
+    if (type == String.class) {
+      AttributeKey<String> key = AttributeKey.stringKey(name);
+      return (setter, arg) -> setter.setAttribute(key, (String) arg);
+    }
+    if (type == long.class || type == Long.class) {
+      AttributeKey<Long> key = AttributeKey.longKey(name);
+      return (setter, arg) -> setter.setAttribute(key, (Long) arg);
+    }
+    if (type == double.class || type == Double.class) {
+      AttributeKey<Double> key = AttributeKey.doubleKey(name);
+      return (setter, arg) -> setter.setAttribute(key, (Double) arg);
+    }
+    if (type == boolean.class || type == Boolean.class) {
+      AttributeKey<Boolean> key = AttributeKey.booleanKey(name);
+      return (setter, arg) -> setter.setAttribute(key, (Boolean) arg);
+    }
+    if (type == int.class || type == Integer.class) {
+      AttributeKey<Long> key = AttributeKey.longKey(name);
+      return (setter, arg) -> setter.setAttribute(key, ((Integer) arg).longValue());
+    }
+    if (type == float.class || type == Float.class) {
+      AttributeKey<Double> key = AttributeKey.doubleKey(name);
+      return (setter, arg) -> setter.setAttribute(key, ((Float) arg).doubleValue());
+    }
+
+    if (isArrayType(type)) {
+      return arrayBinding(name, type);
+    }
+
+    if (isEnumSetType(type)) {
+      return enumSetBinding(name);
+    }
```

Comment:
Probably.  I was trying to handle these cases fairly narrowly to avoid situations where copying the parameter value into the attribute would be expensive.  If we want to support any `Collection<T>` as an attribute of one of the array types that would work, and we could optimize for the cases that don't require copying, e.g. `List<Long>` or `List<String>`.

Reply:
> If we want to support any `Collection<T>` as an attribute of one of the array types that would work, and we could optimize for the cases that don't require copying, e.g. `List<Long>` or `List<String>`.

Hmm, honestly I'm not sure whether we need to support those - I think I'd like to keep it as simple as possible until someone requests that explicitly. Do we need the enum set use case? Is that something you find useful in your code?

Reply:
I nixed EnumSet and related types and kept the array attribute types to only `List<T>`.